### PR TITLE
[6.x] Full PHP 8.0 Support

### PR DESCRIPTION
**Pull request by driesvints**
Draft until most libraries have started to release PHP 8 compatible versions or at least dev builds.

Todo/Requirements:

- [x] `tijsverkoyen/css-to-inline-styles`: https://github.com/tijsverkoyen/CssToInlineStyles/pull/198
- [x] `opis/closure`: https://github.com/opis/closure/pull/67, https://github.com/opis/closure/pull/81
- [x] `guzzlehttp/guzzle`: https://github.com/guzzle/guzzle/issues/2702, https://github.com/guzzle/guzzle/pull/2808
- [x] AWS SDK: https://github.com/aws/aws-sdk-php/pull/2079
- [x] `mockery/mockery` (https://github.com/mockery/mockery/pull/1072, https://github.com/mockery/mockery/pull/1088)
- [x] PHPUnit 9.3 release (will be released on [2020-08-07](https://github.com/sebastianbergmann/phpunit/blob/master/ChangeLog-9.3.md#930---2020-08-07))
- [x] `doctrine/dbal` ^3.0 or ^2.12 release (https://github.com/doctrine/dbal/pull/4347)
- [x] `phpdocumentor/reflection-docblock` ^5.2 release
- [x] Flysystem 1.1 release (https://github.com/thephpleague/mime-type-detection/pull/1, https://github.com/thephpleague/flysystem/pull/1183)
- [x] Re-enable fast-finish
- [x] Prophecy PHP 8 support
- [x] Whoops PHP 8 support (https://github.com/filp/whoops/pull/672)
- [x] Cron expression PHP 8 support (https://github.com/dragonmantank/cron-expression/issues/94)
- [x] Revert extra build steps
- [x] Faker PHP 8 support (https://github.com/FakerPHP/Faker/pull/1)
- [x] Testbench PHP 8 support
- [x] Resolve sort issues

### CODE REVIEW

**Comment by deleugpn** [comment]
I have started the work on opis/closure here: https://github.com/opis/closure/pull/67

File: composer.json
Comment (driesvints) [inline composer.json:85]: @GrahamCampbell why was the caret changed here?

File: src/Illuminate/Database/DPO/MySqlDriver.php
Comment (driesvints) [inline src/Illuminate/Database/DPO/MySqlDriver.php:13]:
```suggestion
            throw new \InvalidArgumentException('Laravel requires the pdo property to be set and be a PDO instance.');
```

File: src/Illuminate/Database/DPO/PostgresDriver.php
Comment (driesvints) [inline src/Illuminate/Database/DPO/PostgresDriver.php:13]:
```suggestion
            throw new \InvalidArgumentException('Laravel requires the pdo property to be set and be a PDO instance.');
```

File: src/Illuminate/Database/DPO/SQLiteDriver.php
Comment (driesvints) [inline src/Illuminate/Database/DPO/SQLiteDriver.php:13]:
```suggestion
            throw new \InvalidArgumentException('Laravel requires the pdo property to be set and be a PDO instance.');
```

File: composer.json
Comment (GrahamCampbell) [reply composer.json:85]: Because we want at least 1.3.2 or at least 1.4.1, but not 1.4.0.

Comment (driesvints) [reply composer.json:85]: Gotcha 👍

File: src/Illuminate/Database/DPO/Connection.php
Comment (deleugpn) [inline src/Illuminate/Database/DPO/Connection.php:3]: Was `DPO` intentional or typo?

Comment (GrahamCampbell) [reply src/Illuminate/Database/DPO/Connection.php:3]: Typo

Comment (driesvints) [reply src/Illuminate/Database/DPO/Connection.php:3]: Fixed

**Comment by GrahamCampbell** [comment]
> Remove phpdocumentor/reflection-docblock from build 

I think this will cause prefer-lowest to fail.

**Comment by driesvints** [comment]
@GrahamCampbell doesn't seems so.

**Comment by GrahamCampbell** [comment]
PHP 8 broke some more stuff, so Mockery is no longer working. https://github.com/mockery/mockery/pull/1088 fixes that.

File: src/Illuminate/Support/Collection.php
Comment (dkulyk) [inline src/Illuminate/Support/Collection.php]: `new static` is missing.

**Review by dkulyk — CHANGES_REQUESTED** [review]

File: src/Illuminate/Support/Collection.php
Comment (GrahamCampbell) [reply src/Illuminate/Support/Collection.php]: This has already been fixed a couple of days ago. You are looking at an outdated diff.

Comment (dkulyk) [reply src/Illuminate/Support/Collection.php]: You are right. Sorry.

File: src/Illuminate/Database/PDO/MySqlDriver.php
Comment (rbuzzibarrios) [inline src/Illuminate/Database/PDO/MySqlDriver.php:10]: If accept my suggestion do not forget include Arr class.
```
    public function connect(array $params)
    {
        $pdoProperty = \Arr::get($params, 'pdo');
        if (! $pdoProperty  instanceof PDO) { // OR if (! ($pdoProperty = \Arr::get($params, 'pdo')) instanceof PDO) {
            throw new \InvalidArgumentException('Laravel requires the pdo property to be set and be a PDO instance.');
        }

        return new Connection($pdoProperty);
    }
```

With this suggestion we can avoid load `if conditions`, avoid call many times `$params['pdo']`.
In addition to, you should create a new Exception for this case, something like `InvalidArgumentPDOException` to avoid repeat 
` throw new \InvalidArgumentException('Laravel requires the pdo property to be set and be a PDO instance.');`
VS
`throw new \InvalidArgumentPDOException();`


On top of that, all the code inside the `connect` method is repeated multiple times, it's nice to encapsulate this method in a single responsibility method and use it however you need it.
helper, interface..

File: src/Illuminate/Database/PDO/PostgresDriver.php
Comment (rbuzzibarrios) [inline src/Illuminate/Database/PDO/PostgresDriver.php:13]:
```suggestion
            throw new \InvalidArgumentPDOException();
```

File: src/Illuminate/Database/PDO/SQLiteDriver.php
Comment (rbuzzibarrios) [inline src/Illuminate/Database/PDO/SQLiteDriver.php:13]:
```suggestion
            throw new \InvalidArgumentPDOException();
```

File: src/Illuminate/Database/Schema/Grammars/RenameColumn.php
Comment (rbuzzibarrios) [inline src/Illuminate/Database/Schema/Grammars/RenameColumn.php:79]:
```suggestion
            return method_exists($column, "set{$name}");
```

File: src/Illuminate/Database/PDO/PostgresDriver.php
Comment (driesvints) [reply src/Illuminate/Database/PDO/PostgresDriver.php:13]: Not necessary.

File: src/Illuminate/Database/Schema/Grammars/RenameColumn.php
Comment (driesvints) [reply src/Illuminate/Database/Schema/Grammars/RenameColumn.php:79]: Unnecessary.

File: src/Illuminate/Database/PDO/MySqlDriver.php
Comment (driesvints) [reply src/Illuminate/Database/PDO/MySqlDriver.php:10]: > With this suggestion we can avoid load if conditions, avoid call many times $params['pdo'].

Unnecessary tbh.

> On top of that, all the code inside the connect method is repeated multiple times

True. I've just pushed a trait that encapsulates this and applied that to the three drivers that were repeating this piece of code. Thanks for the suggestion.

File: src/Illuminate/Database/PDO/SQLiteDriver.php
Comment (driesvints) [reply src/Illuminate/Database/PDO/SQLiteDriver.php]: Unnecessary.

File: src/Illuminate/Support/helpers.php
Comment (GrahamCampbell) [inline src/Illuminate/Support/helpers.php:388]: We can revert this change. It was introduced by a StyleCI regression.

**Review by GrahamCampbell — DISMISSED** [review]

File: src/Illuminate/Support/helpers.php
Comment (driesvints) [reply src/Illuminate/Support/helpers.php]: Done

**Comment by crynobone** [comment]
I added `php-8` branch on orchestra/testbench-core

**Comment by nikic** [comment]
For what it's worth, it would have been really nice to land this PR incrementally. It's okay to have partial PHP 8 support in master with an allow-failures build. It allows other people to start testing.

**Comment by deleugpn** [comment]
@nikic I think the issue would be that there's no 'master' branch for version 6.x and the intention has always been to merge things upwards. This will lend on 6.x and 6.x will be merged into 8.x and master. Merging this on master would make the branches out of sync and harder to resync as we can't merge master (9.x) back into 6.x

**Comment by nikic** [comment]
@deleugpn Sorry for being imprecise there. By "master" I just meant "landed upstream", not necessarily *only* on the master branch. If PHP 8 will be supported on older versions as well, it would of course make sense to land it on lower branches and merge upwards.

For example, Symfony has been landing incremental fixes for PHP 8 support for something like half a year already, across many supported branches, rather than waiting to merge "full support" at once.

**Comment by driesvints** [comment]
@nikic definitely something we can consider. However, I think I'm missing something as I don't see any PHP 8 builds on the https://github.com/symfony/symfony repository?

**Comment by nikic** [comment]
@driesvints It looks like the PHP 8 build is on Travis (nightly): https://travis-ci.org/github/symfony/symfony/builds/736971898

**Comment by driesvints** [comment]
@nikic we've just partially merged this PR and are working on getting the upstream branches up to date. Releases are done tomorrow. Thanks!

**Comment by fridzema** [comment]
https://github.com/doctrine/dbal/releases/tag/2.11.2 🚀

**Comment by ollieread** [comment]
@driesvints Looks like doctrine/dbal#4347 has been merged. Just fzaninotto/Faker#2063 now but it appears that the main dev for that isn't very active with that package.

**Comment by driesvints** [comment]
@ollieread we've contacted the maintainer to move the project to a joint organisation so we can help maintain it. Waiting for a reply now.

**Comment by ollieread** [comment]
Excellent @driesvints , same as I said on the faker pr, should that be required id be happy to help contribute/maintain if there's need.

File: .github/workflows/tests.yml
Comment (driesvints) [inline .github/workflows/tests.yml:57]: Can we get rid off these timeout and attempts in the PR? Don't think we really need those?

Comment (GrahamCampbell) [reply .github/workflows/tests.yml:57]: They are good to stop random failures due to networking. This sometimes is the cause of failing PR tests.

Comment (driesvints) [reply .github/workflows/tests.yml:57]: Okay, we can keep them. No worries.

**Comment by driesvints** [comment]
Putting this in draft again for a bit longer to see if Faker lands PHP 8 support soon.

**Comment by bramceulemans** [comment]
Looks like a fork of Faker is going to have to be made (see https://github.com/fzaninotto/Faker/pull/2063#issuecomment-717118799).

**Comment by driesvints** [comment]
@bramceulemans we saw that and already making plans on how to proceed. Stay tuned.

**Comment by viezel** [comment]
@driesvints I think we need to update the Faker PR to https://github.com/FakerPHP/Faker/pull/1

**Comment by deleugpn** [comment]
This is huge! Congratulations to the Laravel Team!

**Comment by driesvints** [comment]
@deleugpn thanks for your work on this as well 👍

**Pull request merged**
